### PR TITLE
String interning

### PR DESCRIPTION
**Pull request by J1635**
Issue: https://github.com/ACM-UoA-Student-Chapter/dil/issues/6
Continuation of PR: https://github.com/ACM-UoA-Student-Chapter/dil/pull/16
Added test. Edited code according to the comments.

### CODE REVIEW

**Comment by baziotis** [comment]
You dont write "new pull request of...", you just say continuation of and paste the link of the previous ;).
You should not put your code in samples. Samples contain code written on DIL (sample code), _not_ of the compiler.
Also, add more testing. Otherwise, it looks good.

**Comment by J1635** [comment]
Should I create a compiler directory to put my code inside?

**Comment by baziotis** [comment]
Yes, good idea

File: compiler/test_intern_string.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  const char *input_1 = "test";
+  const char *input_2 = "test 2";
+
+  const char *address_1 = intern_string(input_1);
+  const char *address_2 = intern_string(input_2);
+
+  assert(address_1 == intern_string(input_1));
+  assert(address_2 == intern_string(input_2));
+  assert(address_1 != intern_string(input_2));
+  assert(address_2 != intern_string(input_1));
```
Comment (baziotis) [inline compiler/test_intern_string.cpp]: The last 2 are not much useful but the other are good! There's not much to test in here because of most of the functionality is being handled by already tested code (e.g. `strlen`, `vector` etc.). But what other edge cases can you think of?

File: compiler/intern_string.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#include <string>
+#include <vector>
+
+typedef struct InternedStr {
```
Comment (baziotis) [inline compiler/intern_string.cpp]: No need to `typedef` in C++.

**Comment by baziotis** [comment]
> You dont write "new pull request of...", you just say continuation of and paste the link of the previous ;).

Ok that was probably misleading. I meant that you write that in the PR _description_, not in the title.
The title should be just what it does, e.g. Simple string interning

**Comment by baziotis** [comment]
Great and as a final step you can close the other PR.

File: compiler/intern_string.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include <string>
+#include <vector>
+
```
Comment (baziotis) [inline compiler/intern_string.cpp:15]: Add doc about what this module will do.

File: compiler/test_intern_string.cpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  const char *input_2 = "test 2";
+
+  const char *address_1 = intern_string(input_1);
+  const char *address_2 = intern_string(input_2);
+  
+  assert(address_1 != NULL);
+  assert(address_2 != NULL);
+
+  assert(address_1 == intern_string(input_1));
+  assert(address_2 == intern_string(input_2));
```
Comment (baziotis) [inline compiler/test_intern_string.cpp:21]: As I had commented somewhere else, what other edge cases can you think of?

File: compiler/intern_string.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#include <assert.h>
+
+/*
+* intern_string is a function which helps keeping only one copy in memory of each 
+* specific string.
+* intern_string parametre is the string we want to store
+* intern_string return value is a pointer
+* if we have already used the specific string the function returns the old pointer
+* if its a new string we allocate memory and return the new pointer
+*/
```
Comment (baziotis) [inline compiler/intern_string.cpp]: Please use a better format. Something like:
```cpp
/*
 * This helps keep only one copy of every string memory. We pass it a string to
 * be saved. If the string is already saved, we return the old pointer.
 * Otherwise, we save it and return the new pointer. This is called interning a
 * string.
 * Params:
 *    string = The string to be interned.
 * Returns:
 *    A read-only pointer to a copy of `string`.
 */
```

File: compiler/test_intern_string.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#include "intern_string.h"
+#include <assert.h>
+
+int main(int argc, char const *argv[]) {
+  const char *input_1 = "test";
+  const char *input_2 = "test 2";
+  const char *input_3 = "";
+  const char *input_4;
```
Comment (baziotis) [inline compiler/test_intern_string.cpp]: As it seems from the tests, you assume that this is initialized to `NULL` but this is not necessarily the case (that is because it has undefined value - it just _happens_ that most compilers (especially in unoptimized modes) initialize it to 0 for better predictability).
You can initialize it explicitly (i.e. ` = NULL;`).

**Comment by baziotis** [comment]
Don't forget to resolve conversations when you push commit that addresses a review comment. It makes the life of the reviewer much easier ;)

**Comment by baziotis** [comment]
Sorry, I forgot to prompt you to add a license / copyright notice. I added it for you, please verify that it is correct.

**Comment by J1635** [comment]
Yes it's correct.
